hadoop-ozone/ozone-recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestContainerKeyService.java

```diff
@@ -80,41 +92,77 @@
   public void setUp() throws Exception {
     omMetadataManager = initializeNewOmMetadataManager();
-    injector = Guice.createInjector(new AbstractModule() {
-      @Override
-      protected void configure() {
-        try {
-          bind(OzoneConfiguration.class).toInstance(
-              getTestOzoneConfiguration());
-          reconOMMetadataManager = getTestMetadataManager(omMetadataManager);
-          bind(ReconOMMetadataManager.class).toInstance(reconOMMetadataManager);
-          bind(DBStore.class).toProvider(ReconContainerDBProvider.class).
-              in(Singleton.class);
-          bind(ContainerDBServiceProvider.class).to(
-              ContainerDBServiceProviderImpl.class).in(Singleton.class);
-          ozoneManagerServiceProvider = new OzoneManagerServiceProviderImpl(
-              getTestOzoneConfiguration());
-          bind(OzoneManagerServiceProvider.class)
-              .toInstance(ozoneManagerServiceProvider);
-          containerKeyService = new ContainerKeyService();
-          bind(ContainerKeyService.class).toInstance(containerKeyService);
-        } catch (IOException e) {
-          Assert.fail();
+    File tempDir = temporaryFolder.newFolder();
+    AbstractSqlDatabaseTest.DataSourceConfigurationProvider
+        configurationProvider =
+        new AbstractSqlDatabaseTest.DataSourceConfigurationProvider(tempDir);
+
+    JooqPersistenceModule jooqPersistenceModule =
+        new JooqPersistenceModule(configurationProvider);
+
+    injector = Guice.createInjector(jooqPersistenceModule,
+        new AbstractModule() {
+        @Override
+        public void configure() {
+          try {
+            bind(DataSourceConfiguration.class)
+                .toProvider(configurationProvider);
+            OzoneConfiguration configuration = getTestOzoneConfiguration();
+            bind(OzoneConfiguration.class).toInstance(configuration);
+
+            ozoneManagerServiceProvider = new OzoneManagerServiceProviderImpl(
+                configuration);
+
+            reconOMMetadataManager = getTestMetadataManager(omMetadataManager);
+            bind(ReconOMMetadataManager.class)
+                .toInstance(reconOMMetadataManager);
+
+            bind(DBStore.class).toProvider(ReconContainerDBProvider.class).
+                in(Singleton.class);
+            bind(ContainerDBServiceProvider.class)
+                .to(ContainerDBServiceProviderImpl.class).in(Singleton.class);
+
+            bind(OzoneManagerServiceProvider.class)
+                .toInstance(ozoneManagerServiceProvider);
+            containerKeyService = new ContainerKeyService();
+            bind(ContainerKeyService.class).toInstance(containerKeyService);
+          } catch (IOException e) {
+            Assert.fail();
+          }
         }
-      }
-    });
-    containerDbServiceProvider = injector.getInstance(
-        ContainerDBServiceProvider.class);
+      });
+
+    // The following setup is run only once
+    if (!setUpIsDone) {
```

Comment:
@BeforeClass annotated methods are guaranteed to be run once by junit framework.

Reply:
Yes, but @BeforeClass annotated methods need to be static and cannot be used in this case since members of the class need to be updated or accessed.

Reply:
Also, power mock runner does not apply JUnit ClassRules - https://github.com/powermock/powermock/issues/687

The workaround to use `BlockJUnit4ClassRunner.class` also doesn't work as expected and causes a lot of test failures. Hence, sticking with this approach for test classes that use power mock runner.